flatfs.go

```diff
@@ -739,60 +790,100 @@ func (fs *Datastore) updateDiskUsage(path string, add bool) {
 	}
 
 	if fsize != 0 {
-		newDu := atomic.AddInt64(&fs.diskUsage, fsize)
-		fs.checkpointDiskUsage(newDu)
+		atomic.AddInt64(&fs.diskUsage, fsize)
+		fs.checkpointDiskUsage()
 	}
 }
 
-func (fs *Datastore) checkpointDiskUsage(newDuInt int64) {
-	newDu := float64(newDuInt)
-	lastCheckpointDu := float64(atomic.LoadInt64(&fs.diskUsageCheckpoint))
-	diff := math.Abs(newDu - lastCheckpointDu)
-
-	// If the difference between the checkpointed disk usage and
-	// current one is larger than than 1% of the checkpointed: store it.
-	if (lastCheckpointDu * diskUsageCheckpointPercent / 100.0) < diff {
-		fs.persistDiskUsageFile()
+func (fs *Datastore) checkpointDiskUsage() {
+	select {
+	case fs.checkpointCh <- true:
+		// msg sent
+	default:
+		// checkpoint request already pending
 	}
 }
 
-func (fs *Datastore) persistDiskUsageFile() {
-	// Store DiskUsage on clean shutdowns.
-	du, err := fs.DiskUsage()
-	if err != nil {
-		// do not store on errors. just ignore them
-		return
+func (fs *Datastore) checkpointLoop() {
+	timerActive := true
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+	for {
+		select {
+		case _, more := <-fs.checkpointCh:
+			du := atomic.LoadInt64(&fs.diskUsage)
+			fs.dirty = true
+			if !more { // shutting down
+				fs.writeDiskUsageFile(du)
+				if fs.dirty {
+					log.Errorf("could not store final value of disk usage to file, future estimates may be inaccurate")
+				}
+				fs.done <- true
+				return
+			}
+			// If the difference between the checkpointed disk usage and
+			// current one is larger than than `diskUsageCheckpointPercent`
+			// of the checkpointed: store it.
+			newDu := float64(du)
+			lastCheckpointDu := float64(fs.storedValue.diskUsage)
+			diff := math.Abs(newDu - lastCheckpointDu)
+			if (lastCheckpointDu * diskUsageCheckpointPercent / 100.0) < diff {
+				fs.writeDiskUsageFile(du)
+			}
+			// Otherwise insure the value will be written to disk after
+			// `diskUsageCheckpointTimeout`
+			if fs.dirty && !timerActive {
+				timer.Reset(diskUsageCheckpointTimeout)
```

Comment:
We don't appear to ever turn this off.

Reply:
Oops.  Forgot to add `timerActive = true`.

Reply:
Still missing a call to Stop as far as I can tell (and a defered call to Stop).

Reply:
As far as I can tell, there is no need to call Stop here.  If timerActive is false than the timer already expired.  From the go docs (https://golang.org/pkg/time/#Timer.Reset):
> If a program has already received a value from t.C, the timer is known to have expired, and t.Reset can be used directly.

Reply:
Ah. I see.